File: server/src/main/java/com/pass/submission/domain/Answer.java
package com.pass.submission.domain;

import jakarta.persistence.CollectionTable;
import jakarta.persistence.Column;
import jakarta.persistence.ElementCollection;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import java.util.ArrayList;
import java.util.List;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Getter
public class Answer {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private Long questionId;

    @Column
    private String text;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "submission_id")
    private Submission submission;

    @ElementCollection(fetch = FetchType.LAZY)
    @CollectionTable(name = "answer_choice", joinColumns = @JoinColumn(name = "answer_id"))
    private final List<Choice> choices = new ArrayList<>();

    public static Answer textAnswer(Long questionId, String text) {
        return new Answer(questionId, text, new ArrayList<>());
    }

    public static Answer choiceAnswer(Long questionId, List<Choice> choices) {
        return new Answer(questionId, "", choices);
    }

    public Answer(Long questionId, String text, List<Choice> choices) {
        this(null, questionId, text, new ArrayList<>(choices));
    }

    public Answer(Long id, Long questionId, String text, List<Choice> choices) {
        this.id = id;
        this.questionId = questionId;
        this.text = text;
        this.choices.addAll(choices);
    }

    public void updateSubmission(Submission submission) {
        this.submission = submission;
    }
}

File: server/src/main/kotlin/com/fluffy/submission/domain/Answer.kt
package com.fluffy.submission.domain

import jakarta.persistence.CollectionTable
import jakarta.persistence.Column
import jakarta.persistence.ElementCollection
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne

@Entity
class Answer(
    text: String,
    questionId: Long,
    id: Long = 0,
) {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = id
        protected set

    @field:Column(nullable = false)
    var text: String = text
        protected set

    @field:Column(nullable = false)
    var questionId: Long = questionId
        protected set

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "submission_id", nullable = false)
    var submission: Submission? = null
        protected set

    @ElementCollection
    @CollectionTable(name = "answer_choice", joinColumns = [JoinColumn(name = "answer_id")])
    private val _choices: MutableSet<Choice> = mutableSetOf()
    val choices: List<Choice> get() = _choices.toList()

    companion object {

        fun textAnswer(text: String, questionId: Long): Answer {
            return Answer(
                text = text,
                questionId = questionId,
            )
        }

        fun choiceAnswer(questionId: Long, choices: List<Choice>): Answer {
            val answer = Answer(
                text = "",
                questionId = questionId,
            )
            answer.addChoices(choices)

            return answer
        }
    }

    fun addChoices(choices: List<Choice>) {
        _choices.addAll(choices)
    }

    fun updateSubmission(submission: Submission) {
        this.submission = submission
    }
}